### PR TITLE
Properly order multiple emplaced dispatch results.

### LINKED ISSUE #20819: Incorrect concatenation with both Turbine and ONNX compilation
### What happened?

Compiling the following PyTorch module:

```python
import torch
import torch.nn as nn

class IreeBadConcatenation(nn.Module):
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        a = x[:, 1:]
        b = a - a
        c = a + b
        return torch.cat([c, b], dim=1)
```

Produces incorrect results. Namely, the order of `c` and `b` is swapped. The issue does not happen if the tensors are concatenated along the first dimension (`dim=0`) or if the order of `c` and `b` in the concatenation is swapped.

The issue can be reproduced with Turbine and through compilation of imported ONNX model. OnnxRuntime calculates the correct result.

### Steps to reproduce your issue

1. Run the following script:

```python
import subprocess
import sys
import numpy as np
import onnxruntime
import torch
import torch.nn as nn
from iree import compiler as ireec
from iree import runtime as ireert

class IreeBadConcatenation(nn.Module):
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        a = x[:, 1:]
        # Issue is reproducible with other binary operations for `b` and `c` (+, -, *, /)
        b = a - a
        c = a + b
        # Changing the order (`torch.cat([b, c], dim=1)`) or the concatenation
        # dimension (`torch.cat([c, b], dim=0)`) works correctly
        return torch.cat([c, b], dim=1)

if __name__ == "__main__":
    # Test module and example input
    module = IreeBadConcatenation()
    x = np.array([[1.0, 2.0]], dtype=np.float32)

    print("PyTorch")
    print(module(torch.tensor(x)).detach().numpy())

    print("OnnxRuntime")
    # Export
    torch.onnx.export(module, (torch.zeros(x.shape, dtype=torch.float32),), "module.onnx",
                      dynamo=False, input_names=["Input"], output_names=["Output"])
    # Run
    with open("module.onnx", "rb") as f:
        sess = onnxruntime.InferenceSession(f.read())
    print(sess.run(("Output",), {"Input": x})[0])

    print("IREE Turbine")
    turbine_module = torch.compile(module, backend="turbine_cpu")
    print(turbine_module(x))

    print("IREE ONNX")
    # Import
    ret = subprocess.run([sys.executable, "-m", "iree.compiler.tools.import_onnx",
                          "module.onnx", "-o", "module.mlir",])
    ret.check_returncode()
    # Compile
    iree_buffer = ireec.compile_file("module.mlir", optimize=False,
                                     target_backends=["llvm-cpu"], extra_args=["--iree-llvmcpu-target-cpu=generic"])
                                     # Issue is reproducible with VMVX backend
                                     # target_backends=["vmvx"])
    with open("module.vmfb", "wb") as f:
        f.write(iree_buffer)
    # Run
    ireert_config = ireert.Config("local-task")
    ireert_ctx = ireert.SystemContext(config=ireert_config)
    iree_module = ireert.VmModule.copy_buffer(ireert_ctx.instance, iree_buffer)
    ireert_ctx.add_vm_module(iree_module)
    iree_func = ireert_ctx.modules.module["main_graph"]
    print(iree_func(x).to_host())
```

2. Observe the result:

```none
PyTorch
[[2. 0.]]
OnnxRuntime
[[2. 0.]]
IREE Turbine
(...Turbine MLIR...)
[[0. 2.]]
IREE ONNX
[[0. 2.]]
```

### What component(s) does this issue relate to?

Compiler

### Version information

v3.4.0

#### Turbine generated MLIR code:

```mlir
module {
  func.func @main(%arg0: !torch.vtensor<[1,2],f32>) -> !torch.vtensor<[1,2],f32> {
    %int0 = torch.constant.int 0
    %int0_0 = torch.constant.int 0
    %int9223372036854775807 = torch.constant.int 9223372036854775807
    %int1 = torch.constant.int 1
    %0 = torch.aten.slice.Tensor %arg0, %int0, %int0_0, %int9223372036854775807, %int1 : !torch.vtensor<[1,2],f32>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,2],f32>
    %int1_1 = torch.constant.int 1
    %int1_2 = torch.constant.int 1
    %int9223372036854775807_3 = torch.constant.int 9223372036854775807
    %int1_4 = torch.constant.int 1
    %1 = torch.aten.slice.Tensor %0, %int1_1, %int1_2, %int9223372036854775807_3, %int1_4 : !torch.vtensor<[1,2],f32>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1],f32>
    %int1_5 = torch.constant.int 1
    %2 = torch.aten.sub.Tensor %1, %1, %int1_5 : !torch.vtensor<[1,1],f32>, !torch.vtensor<[1,1],f32>, !torch.int -> !torch.vtensor<[1,1],f32>
    %int1_6 = torch.constant.int 1
    %3 = torch.aten.add.Tensor %1, %2, %int1_6 : !torch.vtensor<[1,1],f32>, !torch.vtensor<[1,1],f32>, !torch.int -> !torch.vtensor<[1,1],f32>
    %4 = torch.prim.ListConstruct %3, %2 : (!torch.vtensor<[1,1],f32>, !torch.vtensor<[1,1],f32>) -> !torch.list<vtensor>
    %int1_7 = torch.constant.int 1
    %5 = torch.aten.cat %4, %int1_7 : !torch.list<vtensor>, !torch.int -> !torch.vtensor<[1,2],f32>
    return %5 : !torch.vtensor<[1,2],f32>
  }
}

#map = affine_map<(d0, d1) -> (d0, d1)>
module {
  util.func public @main$async(%arg0: !hal.buffer_view, %arg1: !hal.fence, %arg2: !hal.fence) -> !hal.buffer_view attributes {inlining_policy = #util.inline.never, iree.abi.model = "coarse-fences", iree.abi.stub} {
    %0 = hal.tensor.import wait(%arg1) => %arg0 : !hal.buffer_view -> tensor<1x2xf32>
    %extracted_slice = tensor.extract_slice %0[0, 1] [1, 1] [1, 1] : tensor<1x2xf32> to tensor<1x1xf32>
    %1 = tensor.empty() : tensor<1x1xf32>
    %2 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel", "parallel"]} ins(%extracted_slice, %extracted_slice : tensor<1x1xf32>, tensor<1x1xf32>) outs(%1 : tensor<1x1xf32>) {
    ^bb0(%in: f32, %in_0: f32, %out: f32):
      %6 = arith.subf %in, %in_0 : f32
      linalg.yield %6 : f32
    } -> tensor<1x1xf32>
    %3 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel", "parallel"]} ins(%extracted_slice, %2 : tensor<1x1xf32>, tensor<1x1xf32>) outs(%1 : tensor<1x1xf32>) {
    ^bb0(%in: f32, %in_0: f32, %out: f32):
      %6 = arith.addf %in, %in_0 : f32
      linalg.yield %6 : f32
    } -> tensor<1x1xf32>
    %concat = tensor.concat dim(1) %3, %2 : (tensor<1x1xf32>, tensor<1x1xf32>) -> tensor<1x2xf32>
    %4 = hal.tensor.barrier join(%concat : tensor<1x2xf32>) => %arg2 : !hal.fence
    %5 = hal.tensor.export %4 : tensor<1x2xf32> -> !hal.buffer_view
    util.return %5 : !hal.buffer_view
  }
  util.func public @main(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %0 = util.null : !hal.fence
    %c-1_i32 = arith.constant -1 : i32
    %c0 = arith.constant 0 : index
    %device_0 = hal.devices.get %c0 : !hal.device
    %fence = hal.fence.create device(%device_0 : !hal.device) flags("None") : !hal.fence
    %1 = util.call @main$async(%arg0, %0, %fence) : (!hal.buffer_view, !hal.fence, !hal.fence) -> !hal.buffer_view
    %status = hal.fence.await until([%fence]) timeout_millis(%c-1_i32) flags("None") : i32
    util.return %1 : !hal.buffer_view
  }
}
```

#### ONNX generated MLIR code:

```mlir
module {
  func.func @main_graph(%arg0: !torch.vtensor<[1,2],f32>) -> !torch.vtensor<[1,2],f32> attributes {torch.onnx_meta.ir_version = 8 : si64, torch.onnx_meta.opset_version = 17 : si64, torch.onnx_meta.producer_name = "pytorch", torch.onnx_meta.producer_version = "2.5.1"} {
    %none = torch.constant.none
    %0 = torch.operator "onnx.Constant"() {torch.onnx.value = dense_resource<_> : tensor<1xsi64>} : () -> !torch.vtensor<[1],si64> 
    %1 = torch.operator "onnx.Constant"() {torch.onnx.value = dense_resource<__1> : tensor<1xsi64>} : () -> !torch.vtensor<[1],si64> 
    %2 = torch.operator "onnx.Constant"() {torch.onnx.value = dense_resource<__2> : tensor<1xsi64>} : () -> !torch.vtensor<[1],si64> 
    %3 = torch.operator "onnx.Constant"() {torch.onnx.value = dense_resource<__3> : tensor<1xsi64>} : () -> !torch.vtensor<[1],si64> 
    %4 = torch.operator "onnx.Slice"(%arg0, %1, %2, %0, %3) : (!torch.vtensor<[1,2],f32>, !torch.vtensor<[1],si64>, !torch.vtensor<[1],si64>, !torch.vtensor<[1],si64>, !torch.vtensor<[1],si64>) -> !torch.vtensor<[1,1],f32> 
    %5 = torch.operator "onnx.Sub"(%4, %4) : (!torch.vtensor<[1,1],f32>, !torch.vtensor<[1,1],f32>) -> !torch.vtensor<[1,1],f32> 
    %6 = torch.operator "onnx.Add"(%4, %5) : (!torch.vtensor<[1,1],f32>, !torch.vtensor<[1,1],f32>) -> !torch.vtensor<[1,1],f32> 
    %7 = torch.operator "onnx.Concat"(%6, %5) {torch.onnx.axis = 1 : si64} : (!torch.vtensor<[1,1],f32>, !torch.vtensor<[1,1],f32>) -> !torch.vtensor<[1,2],f32> 
    return %7 : !torch.vtensor<[1,2],f32>
  }
}

{-#
  dialect_resources: {
    builtin: {
      _: "0x080000000100000000000000",
      __1: "0x080000000100000000000000",
      __2: "0x08000000FFFFFFFFFFFFFF7F",
      __3: "0x080000000100000000000000"
    }
  }
#-}
```

### PR DESCRIPTION
**Pull request by benvanik**
#18321 introduced a fix for one issue that caused another: the tied operands added for each result must be in the original result order, but the code was sorting by the order the update ops were present in the block instead. This reworks the pass a bit to support multiple emplace-able results as well as sparsely emplaced results (which prior to this would cause the op to not be emplaced at all). This pass still needs a full rewrite - this is just a quick fix to the correctness issue that was introduced.
The major change here is that all results get turned into tied results from alloca ops and then are selectively replaced by target resource/ranges when possible. This does result in some extra tied operands that don't need to be which is not great for IR size but should result in the same kind of output programs. To fix this all properly an explicit separation of stream bindings between read only, read write, and write only is needed and that is a much larger task.

Fixes #20819.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Stream/Transforms/EmplaceAllocations.cpp [hunk 1]
```diff
@@ -54,74 +64,125 @@ replaceUsesAndTransfer(Value oldValue, Value newValue,
...
+    updateResultOps.emplace_back(updateOp, resultIndex);
+  }
+  if (updateResultOps.empty()) {
+    // No potential updates found, early exit.
+    return false;
+  }
+
+  // Convert the op to be fully emplaced (all results are tied allocas).
+  // We may immediately replace some of these with transfer targets and that's
+  // a waste of IR but we won't know exactly which ones we'll replace until we
```
Comment (nirvedhmeshram) [inline compiler/src/iree/compiler/Dialect/Stream/Transforms/EmplaceAllocations.cpp:145]: I think the comment is unfinished?

**Review by nirvedhmeshram — APPROVED** [review]
LGTM
